In [1]:
import numpy as np
import pandas as pd

In [2]:
data = pd.read_csv("pokemon.csv")

MISSING DATA and TESTING WITH ASSERT

Missing Data: bir veride daha önceden tanımlanmamış, bir değeri olmayan şeylerdir. Bir örnek ile daha iyi anlayalım.

In [3]:
data.head()

#           Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  Sp. Def  \
0  1      Bulbasaur  Grass  Poison  45      49       49       65       65   
1  2        Ivysaur  Grass  Poison  60      62       63       80       80   
2  3       Venusaur  Grass  Poison  80      82       83      100      100   
3  4  Mega Venusaur  Grass  Poison  80     100      123      122      120   
4  5     Charmander   Fire     NaN  39      52       43       60       50   

   Speed  Generation  Legendary  
0     45           1      False  
1     60           1      False  
2     80           1      False  
3     80           1      False  
4     65           1      False

Çıktıda indeksi 4 olan satıra bakın. Type2 değeri NaN'dır. Ve bunun anlamı, Charmander'ın Type2 olarak bir özelliği yok.

Bu şekilde missing data ile karşılaşırsak yapabileceklerimiz şu şekildedir:

- Şu an olduğu gibi bırakabiliriz. Çünkü büyük ihtimal Charmander Type2 değerine shaip değil.
- Missing data 'ları dropna() metodu ile verisetinden çıkarabiliriz. 
- Missing data'ların yerini fillna() metodu ile NaN ile doldurabiliriz.
- Mesela veriler sayısal ise, diğer sample'lere bakarak missing data yerini diğer sample'lerin ortalaması veya medyanı ile doldurabiliriz.

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   #           800 non-null    int64 
 1   Name        799 non-null    object
 2   Type 1      800 non-null    object
 3   Type 2      414 non-null    object
 4   HP          800 non-null    int64 
 5   Attack      800 non-null    int64 
 6   Defense     800 non-null    int64 
 7   Sp. Atk     800 non-null    int64 
 8   Sp. Def     800 non-null    int64 
 9   Speed       800 non-null    int64 
 10  Generation  800 non-null    int64 
 11  Legendary   800 non-null    bool  
dtypes: bool(1), int64(8), object(3)
memory usage: 69.7+ KB


Çıktıyı inceleyelim. Type2 sütununun 414 tane non-null object değeri varmış. Bu durumda Type2 sütununda 386 tane null değer vardır.

In [5]:
data["Type 2"].value_counts(dropna =False)

NaN         386
Flying       97
Ground       35
Poison       34
Psychic      33
Fighting     26
Grass        25
Fairy        23
Steel        22
Dark         20
Dragon       18
Ice          14
Rock         14
Water        14
Ghost        14
Fire         12
Electric      6
Normal        4
Bug           3
Name: Type 2, dtype: int64

Yukarıdaki kodda bulunan value_counts() metodu, Type2'lerindeki her farklı değerden kaç tane olduğunu bulur. Bu metoda parametre olarak yazılan dropna =False ifadesinin anlamı ise, eğer NaN var ise bunu bırakma, onuda göster demek.

Çıktıyı inceleyelim.

Uçan (Flying) 97 tane pokemon olduğunu anlayabiliriz. 

Zehir (Poison) 34 adet olduğunu anlayabiliriz.

Type2 değeri olmayan (NaN) 386 tane pokemon olduğunu anlayabiliriz.

In [6]:
data1=data   
data1["Type 2"].dropna(inplace = True)

Yukarıda yapılan, Type2 sütun değeri NaN olanları "inplace = True" parametresinden dolayı data1'de tutulan verisetinden kalıcı olarak çıkarır.

Yukarıda yaptığımız işlemin işe yarayıp yaramadığını kontrol etmek için "assert" metodu vardır.

In [7]:
assert 1==1

Yukarıdaki ifade sonucunda hiçbir şey dönmüyor ise, doğru demektir.

In [8]:
assert 1==2

AssertionError: 

1, 2'ye eşit olmadığı için hata döndürür.

Şimdi assert metodunu veri setimizde kullanalım.

In [ ]:
assert  data1['Type 2'].notnull().all()

Bir üst satır sonucunda hata almadığımıza göre, Type2 değeri NaN olanları listeden atmışız demektir.

In [ ]:
data["Type 2"].fillna('empty',inplace = True)

Yukarıdaki satırda, Type2'yi empty ile doldurduk.

In [ ]:
assert  data['Type 2'].notnull().all()

Yukarıdaki satır ile tekrar kontrol ettik.

assert ifadesini başka amaçlar içinde kullanabiliriz.

In [ ]:
assert data.columns[1] == 'Name'

Veri setinin 1.sütunu 'Name' mi diye kontrol eder ve doğru olduğundan hiçbir şey döndürmez.

In [ ]:
assert data.Speed.dtypes == np.int

data'nın hızının (Speed) type'si numpy'dan int mi diye kontrol eder. Speed sütununda bulunan değerler int tipinde olduğu için hata döndürmez.